### PR TITLE
#21653 — preparations for switch to Python 3.x

### PR DESCRIPTION
**Pull request by danchr**
Although these changes do not themselves effect the migration to Python 3.5, they constitute the subset of the changes that also apply to Python 2.7.

### CODE REVIEW

**Comment by danchr** [comment]
You might want to look at the changes on a per-commit basis — or I can break them up into smaller chunks.

File: oio_rest/tests/util.py [hunk 1]
```diff
@@ -137,8 +137,9 @@ def setUp(self):
     def tearDown(self):
         super(TestCaseMixin, self).tearDown()
 
-        # for now, we won't restore the settings -- every test should
-        # override them as needed
+        del settings.DB_PORT
+        settings.LOG_AMQP_SERVER = 'localhost'
```
Comment (agger-magenta) [inline oio_rest/tests/util.py]: Det her virker kun, hvis `LOG_AMQP_SERVER` var localhost, før du begyndte at køre testen.

Netop log-tjenesten kunne godt køre op mod en anden LoRa end sig selv p.gr.a. forskellen mellem log og andre tjenester, så det kunne godt være sat til noget andet end localhost.

File: oio_rest/tests/util.py [hunk 2]
```diff
@@ -168,6 +168,8 @@ def assertRequestResponse(self, path, expected, message=None,
             except (IndexError, KeyError, TypeError):
                 pass
 
+        print(r.status_code)
```
Comment (agger-magenta) [inline oio_rest/tests/util.py]: Do we want this in production? I.e., in development?

File: oio_rest/tests/util.py [hunk 3]
```diff
@@ -268,21 +293,24 @@ def assertQueryResponse(self, path, expected, **params):
...
+            path, json=get_fixture(fixture_name), method=method,
+        )
+
+        assert r, 'write of {!r} to {!r} failed!'.format(fixture_name, path)
+
+        objid = r.json.get('uuid')
 
-        self.assertLess(r.status_code, 300)
-        self.assertGreaterEqual(r.status_code, 200)
+        print(r.get_data('True'), path)
```
Comment (agger-magenta) [inline oio_rest/tests/util.py]: Igen, vil vi gerne have den print-statement eller kører vi mere med asserts og "no news is good news"?

File: oio_rest/tests/test_authentication.py [hunk 4]
```diff
@@ -41,3 +44,48 @@ def test_check_saml_authentication_raises_on_invalid_assert(self,
         with self.app.test_request_context(headers=headers), self.assertRaises(
                 AuthorizationFailedException):
             authentication.check_saml_authentication()
+
+
+@patch('oio_rest.settings.USE_SAML_AUTHENTICATION', True)
+class TestVerification(util.TestCase):
```
Comment (cmoesgaard) [inline oio_rest/tests/test_authentication.py]: What are we testing here? What goes wrong during token validation?

File: oio_rest/oio_rest/db.py [hunk 5]
```diff
@@ -90,7 +89,10 @@ def convert_attr_value(attribute_name, attribute_field_name,
     elif field_type == "timestamptz":
         return date_parser.parse(attribute_field_value)
     elif field_type == "interval(0)":
-        return DateTimeDeltaFrom(attribute_field_value).pytimedelta()
+        # delegate actual interval parsing to PostgreSQL in all cases,
+        # bypassing psycopg2 cleverness
+        s = QuotedString(attribute_field_value or '0')
+        return AsIs('{} :: interval'.format(s))
```
Comment (agger-magenta) [inline oio_rest/oio_rest/db.py]: Hvordan virker det her helt nøjagtigt? Vil det sige, at vi ikke kan teste parsning af et interval uden at have Postgres kørende? Det virker måske som en ulempe?

**Comment by agger-magenta** [comment]
I've started a review which probably won't be done till tomorrow.

File: oio_rest/oio_rest/db_helpers.py [hunk 6]
```diff
@@ -54,9 +54,9 @@ def get_attribute_names(class_name):
     "Return the list of all recognized attributes for this class."
     if len(_attribute_names) == 0:
         for c in db_struct:
-            _attribute_names[c] = [
+            _attribute_names[c] = sorted(
```
Comment (agger-magenta) [inline oio_rest/oio_rest/db_helpers.py]: do we have a problem with fluctuations in order? I.e., is it something we see?

If we decide that the attributes should always be sorted this should be part of our future documentation of the format.

File: oio_rest/oio_rest/utils/build_registration.py [hunk 7]
```diff
@@ -205,11 +206,11 @@ def build_registration(class_name, list_args):
...
-    def flatten(d):
-        return [(k, v) for k, v in d.items()]
-
-    all_fields = MultiDict(flatten(states) + flatten(attributes) +
-                           flatten(relations))
+    all_fields = MultiDict(itertools.chain(
+        states.viewitems(),
+        attributes.viewitems(),
+        relations.viewitems(),
+    ))
```
Comment (agger-magenta) [inline oio_rest/oio_rest/utils/build_registration.py]: This is a clear improvement, but does it have a relation to the Python 3 migration or is it just cleanup? (curiosity)

File: oio_rest/tests/util.py [hunk 8]
```diff
@@ -240,7 +240,7 @@ def sort_inner_lists(obj):
             if isinstance(obj, dict):
                 return {
                     k: sort_inner_lists(v)
-                    for k, v in obj.items()
+                    for k, v in obj.iteritems()
```
Comment (agger-magenta) [inline oio_rest/tests/util.py]: Spørgsmålet er, om items vs iteritems giver så meget rent performancemæssigt i de ret små dicts, vi har her. _Men_, det er pænere.

File: oio_rest/tests/test_db.py [hunk 9]
```diff
@@ -1294,6 +1294,7 @@ def test_create_or_import_object_raises_on_pgerror(self, mock_get_conn):
         with self.assertRaises(DBException):
             db.create_or_import_object('', '', '', '')
 
+    @unittest.expectedFailure
```
Comment (agger-magenta) [inline oio_rest/tests/test_db.py]: Man kunne også fjerne denne unittest - giver det ikke mere mening? Nu kaster vi jo ikke længere den exception.

File: oio_rest/tests/test_db_helpers.py [hunk 10]
```diff
@@ -195,6 +197,65 @@ def test_get_attribute_names_reads_db_struct(self):
         # Assert
         self.assertEqual(expected_result, actual_result)
 
+    def test_get_attribute_names_default(self):
+        # Arrange
```
Comment (agger-magenta) [inline oio_rest/tests/test_db_helpers.py]: Disse her tests virker vist ikke med customiserede felter?

File: oio_rest/tests/test_db_helpers.py [hunk 11]
```diff
@@ -559,6 +559,65 @@ def test_get_relation_names_uses_cache(self):
         # Assert
         self.assertEqual(expected_result, actual_result)
 
+    def test_get_state_names_order(self):
+        # Arrange
+        expected_result = {
```
Comment (agger-magenta) [inline oio_rest/tests/test_db_helpers.py]: Is there actually a requirement that state names should be in a particular order? By definition, JSON structures are not ordered.

File: oio_rest/oio_rest/db_helpers.py [hunk 12]
```diff
@@ -65,7 +65,12 @@ def get_attribute_names(class_name):
 
 def get_state_names(class_name):
     "Return the list of all recognized states for this class."
-    return db_struct[class_name.lower()]['tilstande']
+    states = db_struct[class_name.lower()]['tilstande']
+
+    if isinstance(states, list):
```
Comment (agger-magenta) [inline oio_rest/oio_rest/db_helpers.py]: Tilstandene er da altid et dict, er de ikke? I hvert fald i `db_structure.py`.

File: oio_rest/oio_rest/oio_rest.py [hunk 13]
```diff
@@ -62,7 +62,8 @@ def get_virkning_dates(args):
             # current_timestamp,'[]') if possible
             virkning_fra = datetime.datetime.now()
             virkning_til = virkning_fra + datetime.timedelta(
-                    microseconds=1)
+                microseconds=1,
```
Comment (agger-magenta) [inline oio_rest/oio_rest/oio_rest.py]: Trailing commas are ugly,  I think, unless you explicitly use them to create a tuple.

File: oio_rest/oio_rest/db_helpers.py (hunk 6 above)
Comment (danchr) [reply oio_rest/oio_rest/db_helpers.py]: Yes, the attributes are stored in a dictionary, and dictionary order isn't stable in Python. Activity triggers this due to having two states :( I've added a comment to that effect.

File: oio_rest/oio_rest/db_helpers.py (hunk 12 above)
Comment (danchr) [reply oio_rest/oio_rest/db_helpers.py]: Samme commit tilføjer netop nogle lister hvor det er nødvendigt for at bibeholde rækkefølgen.

File: oio_rest/oio_rest/oio_rest.py (hunk 13 above)
Comment (danchr) [reply oio_rest/oio_rest/oio_rest.py]: I've removed the stylistic changes, for now.

File: oio_rest/oio_rest/utils/build_registration.py (hunk 7 above)
Comment (danchr) [reply oio_rest/oio_rest/utils/build_registration.py]: Well, one motivated the other 🙂 I was looking for cases where `2to3` wrapped `items()` in `list()` calls to retain semantics, noticed this one and saw the potential for cleanup.

File: oio_rest/tests/test_authentication.py (hunk 4 above)
Comment (danchr) [reply oio_rest/tests/test_authentication.py]: I've added a (hopefully) helpful docstring.

File: oio_rest/tests/test_db.py (hunk 9 above)
Comment (danchr) [reply oio_rest/tests/test_db.py]: Dropped.

File: oio_rest/tests/util.py (hunk 2 above)
Comment (danchr) [reply oio_rest/tests/util.py]: I like to have a few well-chosen printouts in test code — with a bit of luck, it'll make the simple failure report enough for fixing the bug. Besides, `pytest` supresses these by default.

File: oio_rest/tests/util.py (hunk 8 above)
Comment (danchr) [reply oio_rest/tests/util.py]: It matters for `2to3`, as it attempts to retain semantics by wrapping the inserting `list(x.items())` — which is arguably much worse than either option in Python 2.

File: oio_rest/oio_rest/db.py (hunk 5 above)
Comment (agger-magenta) [reply oio_rest/oio_rest/db.py]: OK!

File: oio_rest/oio_rest/db_helpers.py (hunk 12 above)
Comment (agger-magenta) [reply oio_rest/oio_rest/db_helpers.py]: ok

File: oio_rest/tests/util.py (hunk 2 above)
Comment (agger-magenta) [reply oio_rest/tests/util.py]: ok

**Comment by danchr** [comment]
OK, I've updated the PR so that the tests pass — this required a bit of editing in `tests/util.py` and `mkdb.sh`.

**Pull request merged**